ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -96,14 +101,26 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
    * @param size the number of bytes to upload.
    * @throws CloudStorageException if the upload failed.
    */
-  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size) throws CloudStorageException {
+  private void putBlob(MessageInfo messageInfo, ByteBuffer messageBuf, long size)
+      throws CloudStorageException, IOException, GeneralSecurityException {
     boolean performUpload = messageInfo.getExpirationTimeInMs() == Utils.Infinite_Time && !messageInfo.isDeleted();
-    // May need to encrypt buffer before upload
     if (performUpload) {
       BlobId blobId = (BlobId) messageInfo.getStoreKey();
+      // TODO: would be more efficient to call blobId.isEncrypted()
+      String kmsContext = cryptoService.getEncryptionContext();
+      CloudBlobMetadata.EncryptionType encryptionType;
+      if (cryptoService != null && (blobId.getVersion() < BlobId.BLOB_ID_V4 || !BlobId.isEncrypted(blobId.getID()))) {
+        // Need to encrypt the buffer before upload
+        messageBuf = cryptoService.encrypt(messageBuf);
+        encryptionType = CloudBlobMetadata.EncryptionType.VCR;
```

Comment:
you can add an interface to `CryptoAgent`: `getVcrEncryptionFormat()`,  call it here and pass it to CloudBlobMetadata.

Reply:
vcrEncryptionFormat could be the class name of Encryption Format .